File: ios/Sources/EmarsysPlugin/EmarsysPlugin.swift
import Foundation
import Capacitor
import EmarsysSDK

@objc(EmarsysPlugin)
public class EmarsysPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "EmarsysPlugin"
    public let jsName = "Emarsys"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setContact", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearContact", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackCustomEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setPushToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearPushToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPushToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pauseInApp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resumeInApp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isInAppPaused", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "loadInlineInApp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "changeApplicationCode", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "changeMerchantId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getApplicationCode", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getMerchantId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getClientId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getLanguageCode", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSdkVersion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "enableGeofence", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disableGeofence", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isGeofenceEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getRegisteredGeofences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fetchInboxMessages", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addInboxTag", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeInboxTag", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = EmarsysCore()
    private let push = EmarsysPush()
    private let inApp = EmarsysInApp()
    private let config = EmarsysConfig()
    private let geofence = EmarsysGeofence()
    private let inbox = EmarsysInbox()

    private static let eventName = "emarsysEventHandler"
    private static let inlineEventName = "emarsysInlineInAppHandler"

    // MARK: - Event bus

    /// Called by the Capacitor bridge when the plugin is loaded (after the app's `Emarsys.setup`).
    public override func load() {
        let handler: EMSEventHandlerBlock = { [weak self] name, payload in
            self?.forward(eventName: name, payload: payload)
        }
        Emarsys.push.notificationEventHandler = handler
        Emarsys.push.silentMessageEventHandler = handler
        Emarsys.inApp.eventHandler = handler
        Emarsys.onEventAction.eventHandler = handler
        Emarsys.geofence.eventHandler = handler

        inApp.inlineEventCallback = { [weak self] viewRef, type, data in
            self?.forwardInline(viewRef: viewRef, type: type, data: data)
        }
    }

    private func forward(eventName: String, payload: [AnyHashable: Any]?) {
        var normalizedPayload: [String: Any] = [:]
        if let payload = payload {
            for (key, value) in payload {
                if let key = key as? String {
                    normalizedPayload[key] = value
                }
            }
        }
        let data: [String: Any] = [
            "eventName": eventName,
            "payload": normalizedPayload
        ]
        DispatchQueue.main.async { [weak self] in
            self?.notifyListeners(EmarsysPlugin.eventName, data: data, retainUntilConsumed: true)
        }
    }

    private func forwardInline(viewRef: String, type: String, data: [String: Any]) {
        var payload: [String: Any] = data
        payload["viewRef"] = viewRef
        payload["type"] = type
        DispatchQueue.main.async { [weak self] in
            self?.notifyListeners(EmarsysPlugin.inlineEventName, data: payload)
        }
    }

    // MARK: - Contact

    @objc func setContact(_ call: CAPPluginCall) {
        guard let contactFieldId = call.getInt("contactFieldId") else {
            call.reject("contactFieldId is required")
            return
        }
        guard let contactFieldValue = call.getString("contactFieldValue") else {
            call.reject("contactFieldValue is required")
            return
        }
        
        implementation.setContact(contactFieldId: contactFieldId, contactFieldValue: contactFieldValue) { error in
            if let error = error {
                call.reject("Set contact error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func clearContact(_ call: CAPPluginCall) {
        implementation.clearContact() { error in
            if let error = error {
                call.reject("Clear contact error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func trackCustomEvent(_ call: CAPPluginCall) {
        guard let eventName = call.getString("eventName") else {
            call.reject("eventName is required")
            return
        }
        let eventAttributes = call.getObject("eventAttributes") as NSDictionary? ?? [:]
        
        implementation.trackCustomEvent(eventName: eventName, eventAttributes: eventAttributes) { error in
            if let error = error {
                call.reject("Track custom event error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    // MARK: - Push

    @objc func setPushToken(_ call: CAPPluginCall) {
        guard let pushToken = call.getString("pushToken") else {
            call.reject("pushToken is required")
            return
        }

        push.setPushToken(pushToken: pushToken) { error in
            if let error = error {
                call.reject("Set push token error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func clearPushToken(_ call: CAPPluginCall) {
        push.clearPushToken() { error in
            if let error = error {
                call.reject("Clear push token error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func getPushToken(_ call: CAPPluginCall) {
        let pushToken = push.getPushToken()
        call.resolve([
            "pushToken": pushToken ?? ""
        ])
    }

    // MARK: - InApp

    @objc func pauseInApp(_ call: CAPPluginCall) {
        inApp.pause()
        call.resolve()
    }

    @objc func resumeInApp(_ call: CAPPluginCall) {
        inApp.resume()
        call.resolve()
    }

    @objc func isInAppPaused(_ call: CAPPluginCall) {
        call.resolve(["isPaused": inApp.isPaused()])
    }

    @objc func loadInlineInApp(_ call: CAPPluginCall) {
        guard let viewRef = call.getString("viewRef") else {
            call.reject("viewRef is required")
            return
        }
        guard let viewId = call.getString("viewId") else {
            call.reject("viewId is required")
            return
        }
        guard let frame = frame(from: call) else {
            call.reject("frame is required")
            return
        }
        let zIndex = call.getInt("zIndex")
        DispatchQueue.main.async { [weak self] in
            self?.inApp.loadInline(viewRef: viewRef, viewId: viewId, frame: frame, zIndex: zIndex, webView: self?.webView)
            call.resolve()
        }
    }

    private func frame(from call: CAPPluginCall) -> CGRect? {
        guard let frame = call.getObject("frame") else { return nil }
        let x = (frame["x"] as? NSNumber)?.doubleValue ?? 0
        let y = (frame["y"] as? NSNumber)?.doubleValue ?? 0
        let width = (frame["width"] as? NSNumber)?.doubleValue ?? 0
        let height = (frame["height"] as? NSNumber)?.doubleValue ?? 0
        return CGRect(x: x, y: y, width: width, height: height)
    }

    // MARK: - Config

    @objc func changeApplicationCode(_ call: CAPPluginCall) {
        guard let applicationCode = call.getString("applicationCode") else {
            call.reject("applicationCode is required")
            return
        }
        config.changeApplicationCode(applicationCode) { error in
            if let error = error {
                call.reject("Change application code error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func changeMerchantId(_ call: CAPPluginCall) {
        guard let merchantId = call.getString("merchantId") else {
            call.reject("merchantId is required")
            return
        }
        config.changeMerchantId(merchantId)
        call.resolve()
    }

    @objc func getApplicationCode(_ call: CAPPluginCall) {
        call.resolve(["applicationCode": config.getApplicationCode() ?? ""])
    }

    @objc func getMerchantId(_ call: CAPPluginCall) {
        call.resolve(["merchantId": config.getMerchantId() ?? ""])
    }

    @objc func getClientId(_ call: CAPPluginCall) {
        call.resolve(["clientId": config.getClientId() ?? ""])
    }

    @objc func getLanguageCode(_ call: CAPPluginCall) {
        call.resolve(["languageCode": config.getLanguageCode() ?? ""])
    }

    @objc func getSdkVersion(_ call: CAPPluginCall) {
        call.resolve(["sdkVersion": config.getSdkVersion() ?? ""])
    }

    // MARK: - Geofence

    @objc func enableGeofence(_ call: CAPPluginCall) {
        geofence.enable { error in
            if let error = error {
                call.reject("Enable geofence error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func disableGeofence(_ call: CAPPluginCall) {
        geofence.disable()
        call.resolve()
    }

    @objc func isGeofenceEnabled(_ call: CAPPluginCall) {
        call.resolve(["isEnabled": geofence.isEnabled()])
    }

    @objc func getRegisteredGeofences(_ call: CAPPluginCall) {
        call.resolve(["geofences": geofence.getRegisteredGeofences()])
    }

    // MARK: - Inbox

    @objc func fetchInboxMessages(_ call: CAPPluginCall) {
        inbox.fetchMessages { messages, error in
            if let error = error {
                call.reject("Fetch inbox messages error", error.localizedDescription)
            } else {
                call.resolve(["messages": messages ?? []])
            }
        }
    }

    @objc func addInboxTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("tag is required")
            return
        }
        guard let messageId = call.getString("messageId") else {
            call.reject("messageId is required")
            return
        }
        inbox.addTag(tag, messageId: messageId) { error in
            if let error = error {
                call.reject("Add inbox tag error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func removeInboxTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("tag is required")
            return
        }
        guard let messageId = call.getString("messageId") else {
            call.reject("messageId is required")
            return
        }
        inbox.removeTag(tag, messageId: messageId) { error in
            if let error = error {
                call.reject("Remove inbox tag error", error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }
}

File: android/src/main/java/com/emarsys/plugin/emarsys/EmarsysPlugin.java
package com.emarsys.plugin.emarsys;

import android.graphics.Color;
import android.view.View;
import android.view.ViewGroup;
import com.emarsys.Emarsys;
import com.emarsys.mobileengage.api.event.EventHandler;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "Emarsys")
public class EmarsysPlugin extends Plugin {

    private static final String EVENT_NAME = "emarsysEventHandler";
    private static final String INLINE_EVENT_NAME = "emarsysInlineInAppHandler";

    private EmarsysCore implementation = new EmarsysCore();
    private EmarsysPush push = new EmarsysPush();
    private EmarsysInApp inApp = new EmarsysInApp();
    private EmarsysConfig config = new EmarsysConfig();
    private EmarsysGeofence geofence = new EmarsysGeofence();
    private EmarsysInbox inbox = new EmarsysInbox();

    // Event bus

    /**
     * Called by the Capacitor bridge when the plugin is loaded (after the app's
     * {@code Emarsys.setup}).
     */
    @Override
    public void load() {
        EventHandler handler = (context, eventName, payload) -> forward(eventName, payload);
        Emarsys.getPush().setNotificationEventHandler(handler);
        Emarsys.getPush().setSilentMessageEventHandler(handler);
        Emarsys.getInApp().setEventHandler(handler);
        Emarsys.getOnEventAction().setOnEventActionEventHandler(handler);
        Emarsys.getGeofence().setEventHandler(handler);

        inApp.setInlineEventCallback(this::forwardInline);
    }

    private void forward(String eventName, JSONObject payload) {
        JSObject data = new JSObject();
        data.put("eventName", eventName);
        if (payload != null) {
            try {
                data.put("payload", JSObject.fromJSONObject(payload));
            } catch (JSONException e) {
                data.put("payload", new JSObject());
            }
        } else {
            data.put("payload", new JSObject());
        }
        notifyListeners(EVENT_NAME, data, true);
    }

    private void forwardInline(String viewRef, String type, JSONObject payload) {
        JSObject data = new JSObject();
        data.put("viewRef", viewRef);
        data.put("type", type);
        if (payload != null) {
            Iterator<String> keys = payload.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                data.put(key, payload.opt(key));
            }
        }
        notifyListeners(INLINE_EVENT_NAME, data, false);
    }

    // Contact

    @PluginMethod
    public void setContact(PluginCall call) {
        Integer contactFieldId = call.getInt("contactFieldId");
        String contactFieldValue = call.getString("contactFieldValue");

        if (contactFieldId == null) {
            call.reject("contactFieldId is required");
            return;
        }
        if (contactFieldValue == null) {
            call.reject("contactFieldValue is required");
            return;
        }

        implementation.setContact(contactFieldId, contactFieldValue, (error) -> {
            if (error != null) {
                call.reject("Set contact error", error.getMessage());
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void clearContact(PluginCall call) {
        implementation.clearContact((error) -> {
            if (error != null) {
                call.reject(error.getMessage());
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void trackCustomEvent(PluginCall call) {
        String eventName = call.getString("eventName");

        if (eventName == null || eventName.isEmpty()) {
            call.reject("eventName is required");
            return;
        }

        JSObject attrs = call.getObject("eventAttributes", new JSObject());
        Map<String, String> eventAttributes = new HashMap<>();
        Iterator<String> keys = attrs.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            String value = attrs.optString(key, null);
            if (value != null) {
                eventAttributes.put(key, value);
            }
        }

        implementation.trackCustomEvent(eventName, eventAttributes, (error) -> {
            if (error != null) {
                call.reject("Track custom event error", error.getMessage());
            } else {
                call.resolve();
            }
        });
    }

    // Push

    @PluginMethod
    public void setPushToken(PluginCall call) {
        String pushToken = call.getString("pushToken");

        if (pushToken == null || pushToken.isEmpty()) {
            call.reject("pushToken is required");
            return;
        }

        push.setPushToken(pushToken, (error) -> {
            if (error != null) {
                call.reject(error.getMessage());
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void clearPushToken(PluginCall call) {
        push.clearPushToken((error) -> {
            if (error != null) {
                call.reject(error.getMessage());
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void getPushToken(PluginCall call) {
        String pushToken = push.getPushToken();
        JSObject ret = new JSObject();
        ret.put("pushToken", pushToken != null ? pushToken : "");
        call.resolve(ret);
    }

    // InApp

    @PluginMethod
    public void pauseInApp(PluginCall call) {
        inApp.pause();
        call.resolve();
    }

    @PluginMethod
    public void resumeInApp(PluginCall call) {
        inApp.resume();
        call.resolve();
    }

    @PluginMethod
    public void isInAppPaused(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("isPaused", inApp.isPaused());
        call.resolve(ret);
    }

    @PluginMethod
    public void loadInlineInApp(PluginCall call) {
        String viewRef = call.getString("viewRef");
        String viewId = call.getString("viewId");
        if (viewRef == null) {
            call.reject("viewRef is required");
            return;
        }
        if (viewId == null) {
            call.reject("viewId is required");
            return;
        }
        EmarsysInApp.InlineFrame frame = frameFrom(call);
        if (frame == null) {
            call.reject("frame is required");
            return;
        }
        Integer zIndex = call.getInt("zIndex");
        getActivity().runOnUiThread(() -> {
            ViewGroup parent = webViewParent();
            if (parent == null) {
                call.reject("WebView is not attached");
                return;
            }
            makeWebViewTransparent();
            inApp.loadInline(viewRef, viewId, parent, frame, zIndex);
            call.resolve();
        });
    }

    private EmarsysInApp.InlineFrame frameFrom(PluginCall call) {
        JSObject frame = call.getObject("frame");
        if (frame == null) {
            return null;
        }
        float density = getContext().getResources().getDisplayMetrics().density;
        int x = Math.round((float) frame.optDouble("x", 0) * density);
        int y = Math.round((float) frame.optDouble("y", 0) * density);
        int width = Math.round((float) frame.optDouble("width", 0) * density);
        int height = Math.round((float) frame.optDouble("height", 0) * density);

        View webView = getBridge().getWebView();
        if (webView != null) {
            int[] loc = new int[2];
            webView.getLocationInWindow(loc);
            x += loc[0];
            y += loc[1];
        }
        return new EmarsysInApp.InlineFrame(x, y, width, height);
    }

    private ViewGroup webViewParent() {
        View webView = getBridge().getWebView();
        if (webView != null && webView.getParent() instanceof ViewGroup) {
            return (ViewGroup) webView.getParent();
        }
        return getActivity().findViewById(android.R.id.content);
    }

    private void makeWebViewTransparent() {
        View webView = getBridge().getWebView();
        if (webView != null) {
            webView.setBackgroundColor(Color.TRANSPARENT);
        }
    }

    // Config

    @PluginMethod
    public void changeApplicationCode(PluginCall call) {
        String applicationCode = call.getString("applicationCode");
        if (applicationCode == null) {
            call.reject("applicationCode is required");
            return;
        }
        config.changeApplicationCode(applicationCode, (error) -> {
            if (error != null) {
                call.reject("Change application code error", error.getMessage());
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void changeMerchantId(PluginCall call) {
        String merchantId = call.getString("merchantId");
        if (merchantId == null) {
            call.reject("merchantId is required");
            return;
        }
        config.changeMerchantId(merchantId);
        call.resolve();
    }

    @PluginMethod
    public void getApplicationCode(PluginCall call) {
        JSObject ret = new JSObject();
        String value = config.getApplicationCode();
        ret.put("applicationCode", value != null ? value : "");
        call.resolve(ret);
    }

    @PluginMethod
    public void getMerchantId(PluginCall call) {
        JSObject ret = new JSObject();
        String value = config.getMerchantId();
        ret.put("merchantId", value != null ? value : "");
        call.resolve(ret);
    }

    @PluginMethod
    public void getClientId(PluginCall call) {
        JSObject ret = new JSObject();
        String value = config.getClientId();
        ret.put("clientId", value != null ? value : "");
        call.resolve(ret);
    }

    @PluginMethod
    public void getLanguageCode(PluginCall call) {
        JSObject ret = new JSObject();
        String value = config.getLanguageCode();
        ret.put("languageCode", value != null ? value : "");
        call.resolve(ret);
    }

    @PluginMethod
    public void getSdkVersion(PluginCall call) {
        JSObject ret = new JSObject();
        String value = config.getSdkVersion();
        ret.put("sdkVersion", value != null ? value : "");
        call.resolve(ret);
    }

    // Geofence

    @PluginMethod
    public void enableGeofence(PluginCall call) {
        geofence.enable((error) -> {
            if (error != null) {
                call.reject("Enable geofence error", error.getMessage());
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void disableGeofence(PluginCall call) {
        geofence.disable();
        call.resolve();
    }

    @PluginMethod
    public void isGeofenceEnabled(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("isEnabled", geofence.isEnabled());
        call.resolve(ret);
    }

    @PluginMethod
    public void getRegisteredGeofences(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("geofences", geofence.getRegisteredGeofences());
        call.resolve(ret);
    }

    // Inbox

    @PluginMethod
    public void fetchInboxMessages(PluginCall call) {
        inbox.fetchMessages((messages, error) -> {
            if (error != null) {
                call.reject("Fetch inbox messages error", error.getMessage());
            } else {
                JSObject ret = new JSObject();
                ret.put("messages", messages);
                call.resolve(ret);
            }
        });
    }

    @PluginMethod
    public void addInboxTag(PluginCall call) {
        String tag = call.getString("tag");
        String messageId = call.getString("messageId");
        if (tag == null) {
            call.reject("tag is required");
            return;
        }
        if (messageId == null) {
            call.reject("messageId is required");
            return;
        }
        inbox.addTag(tag, messageId, (error) -> {
            if (error != null) {
                call.reject("Add inbox tag error", error.getMessage());
            } else {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void removeInboxTag(PluginCall call) {
        String tag = call.getString("tag");
        String messageId = call.getString("messageId");
        if (tag == null) {
            call.reject("tag is required");
            return;
        }
        if (messageId == null) {
            call.reject("messageId is required");
            return;
        }
        inbox.removeTag(tag, messageId, (error) -> {
            if (error != null) {
                call.reject("Remove inbox tag error", error.getMessage());
            } else {
                call.resolve();
            }
        });
    }
}
